google/cloud/datastore/query.py

```diff
@@ -246,22 +395,49 @@ def add_filter(self, property_name, operator, value):
                      :class:`google.cloud.datastore.key.Key`
         :param value: The value to filter on.
 
+        :type filter: :class:`CompositeFilter`, :class:`PropertyFiler`
+        :param filter: A instance of a `BaseFilter`, either a `CompositeFilter` or `PropertyFilter`.
+
         :rtype: :class:`~google.cloud.datastore.query.Query`
         :returns: A query object.
 
         :raises: :class:`ValueError` if ``operation`` is not one of the
                  specified values, or if a filter names ``'__key__'`` but
                  passes an invalid value (a key is required).
         """
-        if self.OPERATORS.get(operator) is None:
-            error_message = 'Invalid expression: "%s"' % (operator,)
-            choices_message = "Please use one of: =, <, <=, >, >=, !=, IN, NOT_IN."
-            raise ValueError(error_message, choices_message)
+        if isinstance(property_name, PropertyFilter):
+            raise ValueError(
+                "PropertyFilter object must be passed using keyword argument 'filter'"
+            )
+        if isinstance(property_name, BaseCompositeFilter):
+            raise ValueError(
+                "Or and And objects must be passed using keyword argument 'filter'"
+            )
 
-        if property_name == "__key__" and not isinstance(value, Key):
-            raise ValueError('Invalid key: "%s"' % value)
+        if property_name is not None and operator is not None:
+            if filter is not None:
+                raise ValueError(
+                    "Can't pass in both the positional arguments and 'filter' at the same time"
+                )
+
+            if property_name == "__key__" and not isinstance(value, Key):
```

Comment:
Nit: Can we put `__key__` into a constant somewhere and use that instead? Less risk of accidentally mistyping the number of underscores IMO.

Reply:
Great idea, I've created the `KEY_PROPERTY_NAME` constant and replaced the use of `"__key__"` here.